packages/stylex/src/StyleXSheet.js

```diff
@@ -280,54 +289,123 @@ export class StyleXSheet {
+        );
+        entry.value = newVal;
+        this.constantsCache.set(key, entry);
+        console.log('[defineConsts] cache after updating value:', this.constantsCache);
+
+        for (const idx of entry.rules) {
+          const before = this.rules[idx];
+          this.rules[idx] = before.replaceAll(oldVal, newVal);
+          console.log(
+            `[defineConsts] patched rule index=${idx}`,
+            `before="${before}"`,
+            `after="${this.rules[idx]}"`,
+          );
+        }
+      } else {
+        console.log(`[defineConsts] key=${key} value unchanged (${entry.value}), nothing to do`);
+      }
+      return;
+    }
+
+    // 3) RTL variants
+    if (rawLTRRule != null && rawRTLRule != null) {
       this.insert(
         addAncestorSelector(rawLTRRule, "html:not([dir='rtl'])"),
         priority,
       );
-      this.insert(addAncestorSelector(rawRTLRule, "html[dir='rtl']"), priority);
+      this.insert(
+        addAncestorSelector(rawRTLRule, "html[dir='rtl']"),
+        priority,
+      );
       return;
     }
 
+    // 4) normal rule insertion (stylex.create)
     const rawRule = rawLTRRule;
-
-    // Don't insert this rule if it already exists
     if (this.rules.includes(rawRule)) {
       return;
     }
 
     const rule = this.normalizeRule(
       addSpecificityLevel(rawRule, Math.floor(priority / 1000)),
     );
-
-    // Get the position where we should insert the rule
     const insertPos = this.getInsertPositionForPriority(priority);
     this.rules.splice(insertPos, 0, rule);
-
-    // Set this rule as the end of the priority group
     this.ruleForPriority.set(priority, rule);
+    console.log(`[stylex.create] inserted rule at index=${insertPos}:`, rule);
+    console.log('[stylex.create] cache before scanning for constants:', this.constantsCache);
+
+    // track usage of existing constants in this new rule
+    const CONST_REGEX = /var\(--([A-Za-z0-9_-]+)\)/g;
+    let m;
+    while ((m = CONST_REGEX.exec(rule))) {
+      const key = m[1];
+      const varCall = `var(--${key})`;
+      console.log(`[stylex.create] found ${varCall} in rule index=${insertPos}`);
+
+      let entry = this.constantsCache.get(key);
+      if (!entry) {
+        console.log(`[stylex.create] no existing cache entry for key=${key}, creating one`);
+        entry = { value: null, rules: [] };
+      }
+      entry.rules.push(insertPos);
+      console.log(`[stylex.create] recording rule index=${insertPos} under key=${key}, rules now:`, entry.rules);
+      this.constantsCache.set(key, entry);
+      console.log('[stylex.create] cache after mutation:', this.constantsCache);
+    }
 
+    // 5) finally, insert into the actual CSSOM
     if (this.isHeadless()) {
       return;
     }
-
     const tag = this.getTag();
     const sheet = tag.sheet;
-
     if (sheet != null) {
       try {
         sheet.insertRule(rule, Math.min(insertPos, sheet.cssRules.length));
```

Comment:
When calculating `insertPos`, new rules are "spliced" in and the indices can then move around.

You need to either not rely on indices or make sure you update the indices when you splice things in.